### PR TITLE
feat(stream): add support for xpending command

### LINKED ISSUE #1161: implement xpending

### PR DESCRIPTION
**Pull request by Abhra303**
This PR adds support for `xpending` command. I also created a `StreamCreateNack` function to use test-hookable `GetCurrentTimeMs()` instead of `mstime()`. It gives better control to test stream commands.

Fixes #1161

### CODE REVIEW

File: src/server/stream_family.cc [hunk 1]
```diff
@@ -1116,6 +1128,151 @@ OpResult<uint32_t> OpDel(const OpArgs& op_args, string_view key, absl::Span<stre
...
+  string_view consumer_name;
+  uint64_t delivery_count;
+  mstime_t elapsed;
+};
+
+struct PendingResult {
+  bool is_reduced = true;
+  PendingReducedResult reduced;
+  vector<PendingExtendedResult> extended;
+};
```
Comment (dranikpg) [inline src/server/stream_family.cc:1158]: Such things are usually implemented with a std::variant 

```c++
using PendingExtendedeResultList = std::vector<PendingExtendedResult>;
using PendingResult = std::variant<PendingReducedResult, PendingExtendedeResultList>;
```

File: src/server/stream_family.cc [hunk 2]
```diff
@@ -1116,6 +1128,151 @@ OpResult<uint32_t> OpDel(const OpArgs& op_args, string_view key, absl::Span<stre
...
+  ParsedStreamId end;
+  int64_t min_idle_time = 0;
+  int64_t count = -1;
+};
+
+struct PendingReducedResult {
+  uint64_t count = 0;
+  streamID start;
+  streamID end;
+  vector<pair<string_view, uint64_t>> consumer_list;
```
Comment (dranikpg) [inline src/server/stream_family.cc:1144]: please add a comment what the number is `pair<string_view, uint64_t /* pending? */>`

File: src/server/stream_family.cc [hunk 3]
```diff
@@ -1116,6 +1128,151 @@ OpResult<uint32_t> OpDel(const OpArgs& op_args, string_view key, absl::Span<stre
...
+  raxStart(&ri, cg->consumers);
+  raxSeek(&ri, "^", nullptr, 0);
+  while (raxNext(&ri)) {
+    streamConsumer* consumer = static_cast<streamConsumer*>(ri.data);
+    uint64_t pel_size = raxSize(consumer->pel);
+    if (!pel_size)
+      continue;
+
+    pair<string_view, uint64_t> item;
+    item.first = static_cast<string_view>(consumer->name);
```
Comment (dranikpg) [inline src/server/stream_family.cc:1187]: Eh, not sure 🙂 Looks cool but maybe we also should pass the len with `string_view{name, sdslen(name)}`

**Review by dranikpg — APPROVED** [review]
Left some style comments. Didn't find any critical remarks so I'm gonna approve ahead

File: src/server/stream_family.cc (hunk 2 above)
Comment (Abhra303) [reply src/server/stream_family.cc:1144]: Ah, sure! It is the size of consumer pending list.

File: src/server/stream_family.cc [hunk 4]
```diff
@@ -1151,11 +1151,8 @@ struct PendingExtendedResult {
   mstime_t elapsed;
 };
 
-struct PendingResult {
-  bool is_reduced = true;
-  PendingReducedResult reduced;
-  vector<PendingExtendedResult> extended;
-};
+using PendingExtendedeResultList = std::vector<PendingExtendedResult>;
+using PendingResult = std::variant<PendingReducedResult, PendingExtendedeResultList>;
```
Comment (dranikpg) [inline src/server/stream_family.cc:1155]: You have an e here  PendingExtended**e**ResultList, seems like I added it accidentally when suggesting 😅

File: src/server/stream_family.cc [hunk 5]
```diff
@@ -1730,28 +1726,30 @@ void StreamFamily::XPending(CmdArgList args, ConnectionContext* cntx) {
   }
   PendingResult result = op_result.value();
 
-  if (result.is_reduced) {
-    if (!result.reduced.count) {
+  if (result.index() == 0) {
+    auto res = std::get<PendingReducedResult>(result);
```
Comment (dranikpg) [inline src/server/stream_family.cc:1730]: 1. better to use holds_alternative<>
2. use `const auto& res` to avoid copies

**Review by dranikpg — APPROVED** [review]
:shipit:

**Comment by dranikpg** [comment]
@Abhra303 You can usually merge yourself as soon as someone approved it, but I'm gonna do it now for you 🙂

**Comment by Abhra303** [comment]
I don't have the required permissions for that :)

**Pull request merged**